### ISSUE TITLE
piwigo : sed commands at the end of Dockerfile are not restrictive enough

### ISSUE THREAD

**Issue by jordanauge** [alexbelgium/hassio-addons#38]
The sed command in /etc at the end of the Dockerfile is not restrictive enough and wrongly replaces files like /gallery/include/config_default.inc.php.

**Comment by alexbelgium** [comment]
Hi, thanks for the feedback. Would you have a proposal? I've pushed a new version that searches only in /etc/cont-init.d and /etc/services.d, probably that could suffice

**Comment by alexbelgium** [comment]
It should be corrected, but you need to fully uninstall it first

**Comment by alexbelgium** [comment]
You also need to delete the /config/piwigo folder between two tries if it exists

**Comment by jordanauge** [comment]
Hi, thanks for the quick feedback, and thatnks for the nice work on addons so far. I have made a few quick tries by replacing first config_default and config.inc with known placeholders beforehand but I wanted to try to limit sed to config expressions looking like directories (either followed by space, newline or slash). I have the same issue of PC availability and will be able to look at this from next week :)

**Comment by alexbelgium** [comment]
Hi! Thanks for helping me! The thing is that in HA, only the /data folder is kept when the addon is restarted/updated. That's why I wanted to move the /config folder. I've tried first to move it to /config/piwigo by exposing the /config folder of the host, but in the latest code I've pushed this morning I'm more trying to remove the exposure of /config from host, and instead use /data (which is internal to the addon).

This should prove be easier to manager. However indeed I've still got the error message...

**Comment by alexbelgium** [comment]
Hi, by any chance did you find the solution? Otherwise I'll look again at it. Thanks!

**Comment by alexbelgium** [comment]
Ah, it was easy... There was a remaining line in a script that shouldn't have been there...

**Comment by alexbelgium** [comment]
Can you please confirm if it works? thanks!

**Comment by jordanauge** [comment]
Hello, sorry i could not look at it before. Unless i did something wrong, the local configuration in /gallery/local/config is not preserved across restarts. The only content stored in /data is a dangling symlink due to 

https://github.com/linuxserver/docker-piwigo/blob/master/root/etc/cont-init.d/40-install#L21

![piwigo](https://user-images.githubusercontent.com/7315875/131308488-34a35533-dc91-46d3-bdca-c6d7d3fb14d0.png)

and there are two files of interest for us inside : /gallery/local/config/{config.inc.php, database.inc.php}.

- database.inc.php stores the install settings, including database configuration and admin account
- config.inc.php is initialized from the default settings, and has an additional issue we should fix :

```
$conf['show_php_errors'] = E_ALL;
```

This triggers warnings like the following, which breaks the visual appearance of the website, and more importants the parsing of all RPC calls within piwigo (seen with a wrong json format):

```
 Deprecated: Invalid characters passed for attempted conversion, these have been ignored in /gallery/include/template.class.php on line 1024
```

This should be the correct setting:

$conf['show_php_errors'] = E_ALL & ~E_NOTICE & ~E_DEPRECATED;

I'm still not fluent with docker and its use within home assistant but i hope this is already useful.

**Comment by jordanauge** [comment]
We would need a way to change the path to the local config which is typically built like this : PHPWG_ROOT_PATH.PWG_LOCAL_DIR. Any idea ?

**Comment by alexbelgium** [comment]
it should be done easily with a sed command, I'll look at it

**Comment by alexbelgium** [comment]
or an ENV variable for PHPWG_ROOT_PATH :-) I'll test

**Comment by alexbelgium** [comment]
doesn't seem to work with an ENV variable

**Comment by alexbelgium** [comment]
Could you please test for remanence? I don't have a test database. Normally it should work, I've symlinked the /gallery/local/config and /config/keys to the /data folder. Thanks!

**Comment by alexbelgium** [comment]
the config will be exposed in /share/piwigo/config and /share/piwigo/keys

**Comment by alexbelgium** [comment]
and I've added a line to correct the config if needed in the 41-folder script

**Comment by alexbelgium** [comment]
no, it doesn't work yet

**Comment by alexbelgium** [comment]
New version, it should work (including config file correction)

**Comment by jordanauge** [comment]
All working indeed, thanks and great work !

**Comment by alexbelgium** [comment]
That's perfect, thanks for the great help identifying how to solve the issues! It was very useful in finding the solution.

**Comment by jordanauge** [comment]
Do you think it is worth reporting the issue upstream to linuxserver.io, as I believe their setup does not allow to preserve configuration across restarts ?

**Comment by alexbelgium** [comment]
Actually it's purely due to how HA works! Each time you restart the addon, it recreates a container based on the initial image. So every change is wiped clean, except on the mapped volume which is by default /data. This causes the lack of rémanence. It is the possible to map other volumes (/config. /ssl, /media or /share for example), but bottom line to have remanence in an HA addon we need to modify the upstream image... I hope this was clear... 

Bottom line this is what I do in my addon : create a shell around a preexisting image to map HA addon options with the image options, modify files to allow rémanence,plus add a few scripts like mounting local drives (which is the also not allowed by default in HA)...
